### ISSUE TITLE
option to have git completion show only local branches as shown by git branch -v

### ISSUE THREAD

**Issue by greggman** [ohmyzsh/ohmyzsh#11132]
### If the feature request is for a plugin or theme, specify it here.

git plugin

### If the feature solves a problem you have, specify it here.

The problem I'm running into is git completion for `git checkout` and maybe other commands shows too many completions.

Example: `git branch -v` shows just 4 branches

```
> git branch -v
  dangling-pointer       ce141cd55 Metal: Fix dangling pointer in IntermRebuild
* fix-occlusion-query-m1 67a147feb Metal: Fix Occlusion Query bug on M1
  main                   93c0c90db Suppress the failing egl test to unblock roller
  readpixel-speed-test   3ebef5614 Metal:Read Pixels Speed Test
```

But `git checkout <tab>` shows

```
❯ git checkout <tab>
zsh: do you wish to see all 6036 possibilities (776 lines)? 
```

Even `git checkout --no-guess <tab>` shows

```
> git checkout --no-guess <tab>
zsh: do you wish to see all 3298 possibilities (1113 lines)? 
```

which is 3294 more than is useful to me. `export GIT_COMPLETION_CHECKOUT_NO_GUESS=1` seems to do nothing

```
❯ export GIT_COMPLETION_CHECKOUT_NO_GUESS=1
❯ git checkout <tab>
zsh: do you wish to see all 6036 possibilities (776 lines)? 
```

### Describe the proposed feature.

I'm guessing it would be a tiny changes to add something like `GIT_COMPLETION_CHECKOUT_LOCAL_ONLY` had have it filter the same way `git branch` filters

### Describe alternatives you've considered

Not really sure what else to consider. Even if I copy `.git-completion.bash` locally I'm not sure what to change and I suspect others would find this feature useful.

**Comment by brokenalarms** [comment]
yeah, this is a longstanding issue generally with git completions. A guide to do it manually is here: https://cmetcalfe.ca/blog/git-checkout-autocomplete-local-branches-only.html

But it looked like `gitfast` changed this code in `git-completion.bash` to begin with, so you can't search for the same thing.

This was the first optimization I expected `gitfast` to do, so a little surprised it's not included :)

**Comment by robbyrussell** [comment]
Hey @greggman, sorry — nobody ever replied to this. That's on us.

The short version: that completion isn't ours. Oh My Zsh doesn't ship a `_git` outside the opt-in `gitfast` plugin, and we set no git completion styles at all — the list you're seeing comes from zsh's own `_git`. So it's configurable in your `.zshrc` without any change here.

This works, and I tested it on zsh 5.9:

```zsh
zstyle ':completion:*:*:git-checkout:*' tag-order '! heads-remote remote-branch-names-noprefix commit-tags commit-objects commits recent-branches'
```

Two things that tripped me up and are probably worth saying, since they're why this looks broken:

- The intuitive version, `tag-order 'heads-local'`, does **not** work. zsh falls back to offering every tag when the requested one isn't in the outer tag set, so you get more, not less.
- If your `git` is Homebrew's, you may be getting git's own completion rather than zsh's, and no zstyle affects that one. `GIT_COMPLETION_CHECKOUT_NO_GUESS=1` only removes the DWIM shorthand names, not `origin/*` or tags — which is likely why that seemed to do nothing.

Marc looked at the same request back in #10323 and decided it was too niche to build into the git plugin, and I think that still holds. Closing this as the same thing, but hopefully the snippet above is what you actually wanted.

🤖 Co-authored with Claude Code

**Comment by greggman** [comment]
Wow! no worries. Thanks so much for looking into it! 🙇‍♂️
